### ISSUE TITLE
Validation bug: InvokeArgumentValidator when validating DynamicInvokeExpr

### ISSUE THREAD

**Issue by xenox1886** [Sable/soot#976]
When validating a method body with the `InvokeArgumentValidator` (manually or when writing a class file with Jasmin), the `validate` method calls `getMethod` on the `InvokeExpr` of the `Stmt` currently being checked:
```java
InvokeExpr iinvExpr = s.getInvokeExpr();
SootMethod callee = iinvExpr.getMethod();
```
However, if `iinvExpr` happens to be of type `DynamicInvokeExpr`, and therefore not resolvable at compile time, `SootMethodRefImpl.resolve` (which gets called by `iinvExpr.getMethod`) will throw a `ClassResolutionFailedException`. 

This is particularly annoying, because `InvokeArgumentValidator` is a basic validator, i.e. (afaik) disabling soot's validation option doesn't make a difference, since it gets called anyway.
It also entails that one can't read dynamic expressions (e.g. when using lambdas) and writing them, without any changes, back with Jasmin, since the validation fails.

Possible (quick and dirty) fix in `InvokeArgumentValidator.validate`: 
```java
InvokeExpr iinvExpr = s.getInvokeExpr();
if (iinvExpr instanceof DynamicInvokeExpr){
    continue;
}
SootMethod callee = iinvExpr.getMethod();
```

(Also I assume the first line of `SootMethodRefImpl.tryResove` should actually throw an exception in this case, but it fails to compare the class name to "soot.dummy.InvokeDynamic")

**Comment by mbenz89** [comment]
I find it a bit strange that this problem wasn't observed before. Having lambdas in client code (and thus DynamicInvokes) shouldn't be a corner case.
Is there a reason for which you are using Jasmin? I am pretty sure that Soot defaults to the ASM Backend in the current version.

Can you recheck with ASM and if it still faile propose a PR with your fix(es)?

**Comment by xenox1886** [comment]
No, I have no particular reason to use Jasmin, I just followed this tutorial: https://github.com/Sable/soot/wiki/Creating-a-class-from-scratch
I've switched to the BafASM backend now (that's what you meant, right?), but I get the same error (see stacktrace below). I'll make a PR, as soon as I got the time.
Thanks
```java
 Exception in thread "main" soot.SootMethodRefImpl$ClassResolutionFailedException: Class soot.dummy.InvokeDynamic doesn't have method accept([]) : java.util.function.Consumer; failed to resolve in superclasses and interfacesLooking in soot.dummy.InvokeDynamic which has methods []

	at soot.SootMethodRefImpl.resolve(SootMethodRefImpl.java:226)
	at soot.SootMethodRefImpl.resolve(SootMethodRefImpl.java:148)
	at soot.jimple.internal.AbstractInvokeExpr.getMethod(AbstractInvokeExpr.java:67)
	at soot.jimple.validation.InvokeArgumentValidator.validate(InvokeArgumentValidator.java:32)
	at soot.jimple.JimpleBody.validate(JimpleBody.java:121)
	at soot.jimple.JimpleBody.validate(JimpleBody.java:103)
	at soot.baf.BafBody.<init>(BafBody.java:91)
	at soot.baf.Baf.newBody(Baf.java:560)
	at soot.PackManager.convertJimpleBodyToBaf(PackManager.java:1064)
	at soot.AbstractASMBackend.getBafBody(AbstractASMBackend.java:141)
	at soot.baf.BafASMBackend.getMinJavaVersion(BafASMBackend.java:110)
	at soot.AbstractASMBackend.getMinJavaVersion(AbstractASMBackend.java:199)
	at soot.AbstractASMBackend.<init>(AbstractASMBackend.java:85)
	at soot.baf.BafASMBackend.<init>(BafASMBackend.java:100)
```

**Comment by mbenz89** [comment]
Yep, that's what I meant. I suggest you stay with ASM since Jasmin is deprecated and not supported anymore. Sorry for the outdated tutorials.

Just to be sure, could you execute `PackManager.writeOutput()` which should output all application classes with the configured output options.

**Comment by xenox1886** [comment]
No problem.
Ok, that's the resulting stack trace:
```java
Exception in thread "main" soot.SootMethodRefImpl$ClassResolutionFailedException: Class soot.dummy.InvokeDynamic doesn't have method accept([]) : java.util.function.Consumer; failed to resolve in superclasses and interfacesLooking in soot.dummy.InvokeDynamic which has methods []

	at soot.SootMethodRefImpl.resolve(SootMethodRefImpl.java:226)
	at soot.SootMethodRefImpl.resolve(SootMethodRefImpl.java:148)
	at soot.jimple.internal.AbstractInvokeExpr.getMethod(AbstractInvokeExpr.java:67)
	at soot.jimple.validation.InvokeArgumentValidator.validate(InvokeArgumentValidator.java:32)
	at soot.jimple.JimpleBody.validate(JimpleBody.java:121)
	at soot.jimple.JimpleBody.validate(JimpleBody.java:103)
	at soot.baf.BafBody.<init>(BafBody.java:91)
	at soot.baf.Baf.newBody(Baf.java:560)
	at soot.PackManager.convertJimpleBodyToBaf(PackManager.java:1064)
	at soot.AbstractASMBackend.getBafBody(AbstractASMBackend.java:141)
	at soot.baf.BafASMBackend.getMinJavaVersion(BafASMBackend.java:110)
	at soot.AbstractASMBackend.getMinJavaVersion(AbstractASMBackend.java:199)
	at soot.AbstractASMBackend.<init>(AbstractASMBackend.java:85)
	at soot.baf.BafASMBackend.<init>(BafASMBackend.java:100)
	at soot.PackManager.writeClass(PackManager.java:1134)
	at soot.PackManager$2.run(PackManager.java:693)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

In the meantime, I've tried writing the class with phantom refs enabled and that works. I guess this has something to to with this (maybe dated?) article, I've stumbled upon (https://github.com/Sable/soot/wiki/Java%27s-Invokedynamic) containing the interesting sentence "In this signature, the name of the declaring class must be soot.dummy.InvokeDynamic. Soot will handle this special class as a phantom class (even if phantom classes are disabled otherwise.)". However, I didn't find special treatment of `soot.dummy.InvokeDynamic` as a phantom class, so that may be the actual issue here. I'll try to fix that in the PR.

**Comment by xenox1886** [comment]
PR #979

**Comment by mbenz89** [comment]
Fixed with #979
